### PR TITLE
Feature/print struct like objects

### PR DESCRIPTION
**Pull request by waldyr**
#168

### CODE REVIEW

**Comment by gerrywastaken** [comment]
Thank you for the submission and sorry about the long delay in response. The failures in our Travis builds on master have now been fixed (`git rebase upstream/master` to update your commits) and I'm trying to go through outstanding PRs when I get time.

I tried rebasing this on master using my fork, but ended up with errors in the build: https://travis-ci.org/gerrywastaken/awesome_print/builds/129044884

**Comment by waldyr** [comment]
> Thank you for the submission and sorry about the long delay in response.

No worries at all. Time is an issue for everyone 😄 

> [...] I'm trying to go through outstanding PRs when I get time.

I'm really glad that you liked it even though I think that there should be a better solution. In that case I want to apologize for adding more cyclomatic complexity (that one conditional to check if it's an ivar or method). Maybe you can come up with something more elegant and I would be glad if you let me know about that.

I wish I could decouple some of the logic from the formatter's methods into specific classes for each handled type but as I said time is an issue for everyone.

File: lib/awesome_print/formatter.rb [hunk 1]
```diff
@@ -158,7 +158,15 @@ def awesome_object_data(o, variables)
...
+          result = colorize(" = ", :hash)
+
+          if var.to_s.start_with?('@')
+            result += @inspector.awesome(o.instance_variable_get(var))
+          else
+            result += @inspector.awesome(o.send(var))
+          end
+
+          key << result
         end
```
Comment (gerrywastaken) [inline lib/awesome_print/formatter.rb]: How do you feel about the following: 

* Changing the `if` condition to a call, to make it a bit more descriptive and pulling the logic outside of this method?
* We could also write `result += @inspector.awesome(` after the branching to save writing it twice.
* Adding a commit to the send line to explain it's inclusion.

e.g.
```ruby
private

def valid_instance_var?(variable_name)
  variable_name.to_s.start_with?('@')
end

# [...]

  indented do
    var_contents = if valid_instance_var?(var)
      o.instance_variable_get(var)
    else
      o.send(var) # Enables handling of Struct attributes
    end

    key << colorize(" = ", :hash) + @inspector.awesome(var_contents)
  end
```

Thoughts?

**Comment by gerrywastaken** [comment]
> In that case I want to apologize for adding more cyclomatic complexity (that one conditional to check if it's an ivar or method). 

That method was already doing way too much, it needs a complete overhaul.

> I wish I could decouple some of the logic from the formatter's methods into specific classes

Agreed, that would be good in the long term.

I've added some comments to your last commit.

File: lib/awesome_print/formatter.rb (hunk 1 above)
Comment (waldyr) [inline lib/awesome_print/formatter.rb]: That's way better than before. Tonight I'll patch it up.

**Comment by waldyr** [comment]
Now that I can rely on the CI, next PR I'm gonna work on decoupling those methods into new classes. What do you think?

**Comment by gerrywastaken** [comment]
Sounds fantastic! Although pease break it up into logical commits, like you have done with this one. :) That way it will be easy for me and everybody else to follow the changes.

Thanks heaps for this PR. 👍  💯 

This will be in the next release, however I'm still waiting on access atm: https://github.com/michaeldv/awesome_print/issues/234

**Comment by gerrywastaken** [comment]
This change went like in version 1.7.0. 👍

**Pull request merged**